pravega/pravega
# Issue 3411: Improve permissions required for Controller gRPC API

**Change log description**  

Some of the operations of the Controller gRPC interface require a higher permission level `READ_UPDATE`, even though they are read operations and do not modify the state of the resource. And then, there is one operation that does modify the state, but only requires `READ` permission. 

**Purpose of the change**  
Fixes #3411. 

**What the code does**  

The code fixes the permissions required for the operations listed below. We continue to use the same serial numbers (S.No.) as mentioned in the issue so that it is easier for readers to keep track of the changes. 

|S.No.| Action |	Resource String for Auth (in this PR)  | Required Permission - "OLD" | Required Permission - "NEW"| gRPC method (in ``ControllerServiceImpl``) | REST controller method (in ``StreamMetadataResourceImpl``| 
|:--:|:-----------|:---------------:|:---------------:|:---------------:|:---------------:|:----------------------|
|21.| Check scale status | ``<scope-name>/<stream-name>`` | READ_UPDATE | READ | ``checkScale``|-|
|22.| Get URI for segment | ``<scope-name>/<stream-name>`` | READ_UPDATE | READ | ``getURI``|-|
|23.| Is Segment valid | ``<scope-name>/<stream-name>`` | READ_UPDATE | READ | ``isSegmentValid``|-|
|24.| Is stream cut valid | ``<scope-name>/<stream-name>`` | READ_UPDATE | READ | ``isStreamCutValid``|-|
|28.| Ping transaction | ``<scope-name>/<stream-name>`` | READ | READ_UPDATE | ``pingTransaction``|-|
|29.| Check transaction state | ``<scope-name>/<stream-name>`` | READ_UPDATE | READ | ``checkTransactionState``|-|  

Also, a new test class `ControllerGrpcAuthFocusedTest` has been added for verifying authorization logic of the Controller gRPC service implementation `ControllerServiceImpl`.

**Note:**
* The permission required for `Get delegation token`/`getDelegationToken` has been left as-is (`READ_UPDATE`) as the operation is state-mutative.
* The permission required for `List current stream segments`/`getCurrentSegments()` and `List stream segments`/`getSegments()` have been left as-is (`READ_UPDATE`) as the operations use the permission to create a delegation token for downstream processing.

**How to verify it**  

All tests should pass. 

For convenience, I also create a temporary test suite comprising of a subset of test classes as shown below, for verifying code when there are any auth-related changes in the controller interfaces:

```java
import io.pravega.controller.rest.v1.SecureStreamMetaDataTests;
import io.pravega.controller.rest.v1.StreamMetaDataAuthFocusedTests;
import io.pravega.controller.rest.v1.StreamMetaDataTests;
import io.pravega.controller.rest.v1.UserSecureStreamMetaDataTests;
import io.pravega.controller.server.v1.InMemoryControllerServiceImplTest;
import io.pravega.controller.server.v1.ZKControllerServiceImplTest;
import org.junit.runner.RunWith;
import org.junit.runners.Suite;

@RunWith(Suite.class)
@Suite.SuiteClasses({
        StreamMetaDataAuthFocusedTests.class,
        StreamMetaDataTests.class,
        SecureStreamMetaDataTests.class,
        UserSecureStreamMetaDataTests.class,
        InMemoryControllerServiceImplTest.class,
        ZKControllerServiceImplTest.class,
        RESTAuthHelperTest.class,
        PravegaAuthManagerTest.class,
        ControllerGrpcAuthFocusedTest.class
})

// Run this class on your IDE to execute the test classes in the suite. 
public class AuthTestSuite {
}
```
